### PR TITLE
feat: add support for quantized-llama using GGML

### PR DESCRIPTION
**Pull request by grepfruit19**
Can test by modifying quickstart.py to be the following:

`client = Client(big_model='quantized-llama', small_model='quantized-llama')`

### CODE REVIEW

**Comment by grepfruit19** [comment]
closes #136

File: src/llm_vm/onsite_llm.py [hunk 1]
```diff
@@ -303,6 +304,36 @@ def model_loader(self):
...
+    Methods:
+        model_loader: Loads the LLM into memory
+        tokenizer_loader: Does nothing. CTransformers includes its own tokenizer so this is unnecessary
+        generate: Generates a response from a given prompt with the loaded LLM and tokenizer
+    """
+    model_uri="TheBloke/LLaMa-7B-GGML"
+
+    def model_loader(self):
+        # This file specifically is the smallest model.
+        return AutoModelForCausalLM.from_pretrained(self.model_uri, model_file="llama-7b.ggmlv3.q2_K.bin")
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:326]: question: why are you specifying both file and uri?

File: src/llm_vm/onsite_llm.py [hunk 2]
```diff
@@ -303,6 +304,36 @@ def model_loader(self):
...
+
+    # CTransformers loads its tokenizer in the model, so this function is unnecessary
+    def tokenizer_loader(self):
+        return
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:330]: no, we still need to initialize the tokenizing/detokenizing interface right? or does the stubs get set correctly?

File: src/llm_vm/onsite_llm.py [hunk 3]
```diff
@@ -303,6 +304,36 @@ def model_loader(self):
...
+
+    def generate(self, prompt, max_length=100, **kwargs):
+        inputs = self.model.tokenize(prompt)
+        generated_tokens = self.model.generate(inputs)
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:334]: shouldnt we pass in kwargs? i think i have some example where we do that, or is that still a massive TODO fixme?

File: src/llm_vm/onsite_llm.py (hunk 1 above)
Comment (cartazio) [reply src/llm_vm/onsite_llm.py:326]: or is it file in the repo?

File: src/llm_vm/onsite_llm.py [hunk 4]
```diff
@@ -303,6 +304,36 @@ def model_loader(self):
...
     def tokenizer_loader(self):
         return LlamaTokenizer.from_pretrained(self.model_uri)
+    
+@RegisterModelClass("quantized-llama")
+class Quantized_Llama(Base_Onsite_LLM):
+    """
+    Class for running quantized Llama instances that use GGML
+
+    Attributes:
+        model_uri (str): Hugging Face Endpoint for LLM
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:314]: are we gonna need model_file_name for this?
how does accessing a local model work vs remote model?
with the others, the URI could be a local file path i think, though its probably not very robust

File: src/llm_vm/onsite_llm.py (hunk 1 above)
Comment (grepfruit19) [reply src/llm_vm/onsite_llm.py:326]: There are several versions of the model, with varying quantizations. The one I listed is the 2bit model, I just used the smallest model they have.

Comment (cartazio) [reply src/llm_vm/onsite_llm.py:326]: i guess ggml repos tend to have more than one model to pick from
ok cool, sounds good

Comment (cartazio) [reply src/llm_vm/onsite_llm.py:326]: we do need some way to overrride model file name though, i think you need to setup an __init__ that takes the right kwargs with default values so you can thread stuff into that space

File: src/llm_vm/onsite_llm.py (hunk 3 above)
Comment (grepfruit19) [reply src/llm_vm/onsite_llm.py:334]: I'll add kwargs, some classes like the GPT ones take kwargs but not sure how fleshed out those implementations are

File: src/llm_vm/onsite_llm.py (hunk 1 above)
Comment (grepfruit19) [reply src/llm_vm/onsite_llm.py:326]: ^ working on that now

File: src/llm_vm/onsite_llm.py (hunk 4 above)
Comment (grepfruit19) [reply src/llm_vm/onsite_llm.py:314]: It looks like ctransformers behaves the same way: https://github.com/marella/ctransformers#classmethod-automodelforcausallmfrom_pretrained

File: src/llm_vm/onsite_llm.py (hunk 2 above)
Comment (grepfruit19) [reply src/llm_vm/onsite_llm.py:343]: I _think_ ctransformers abstracts the Tokenizer away. I put up an issue asking more details on their github.

The generate function handles its own tokenizing/detokenizing and it seems to work properly, though I'm not sure if that Tokenizer is used elsewhere (I couldn't find anywhere that it is).

File: src/llm_vm/onsite_llm.py [hunk 5]
```diff
@@ -303,6 +304,49 @@ def model_loader(self):
...
+        __init__: Takes the same parameters as the base class, except it looks to see if there is a specific quantization you want to use
+        Quantized weights can be found (here)[https://huggingface.co/TheBloke/LLaMa-7B-GGML#provided-files]
+        model_loader: Loads the LLM into memory
+        tokenizer_loader: Does nothing. CTransformers includes its own tokenizer so this is unnecessary
+        generate: Generates a response from a given prompt with the loaded LLM and tokenizer
+    """
+
+    model_uri="TheBloke/LLaMa-7B-GGML"
+
+    def __init__(self,model_uri=None,tokenizer_kw_args={},model_kw_args={}):
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:327]: dont you need to set / get model_uri and model file as KWargs?

File: src/llm_vm/onsite_llm.py [hunk 6]
```diff
@@ -303,6 +304,49 @@ def model_loader(self):
...
+
+    Methods:
+        __init__: Takes the same parameters as the base class, except it looks to see if there is a specific quantization you want to use
+        Quantized weights can be found (here)[https://huggingface.co/TheBloke/LLaMa-7B-GGML#provided-files]
+        model_loader: Loads the LLM into memory
+        tokenizer_loader: Does nothing. CTransformers includes its own tokenizer so this is unnecessary
+        generate: Generates a response from a given prompt with the loaded LLM and tokenizer
+    """
+
+    model_uri="TheBloke/LLaMa-7B-GGML"
```
Comment (cartazio) [inline src/llm_vm/onsite_llm.py:325]: shoudln't this be a default kwarg?

**Comment by cartazio** [comment]
Rebase it!

**Comment by cartazio** [comment]
@grepfruit19  could you rebase + move the example script into the examples folder? 
thx :)

**Comment by grepfruit19** [comment]
@cartazio Just rebased!

File: src/llm_vm/onsite_llm.py (hunk 5 above)
Comment (cartazio) [reply src/llm_vm/onsite_llm.py:329]: You still need to set the default kwarg here

**Comment by cartazio** [comment]
Great. Let’s populate those default kwargs and get crackin!

**Comment by cartazio** [comment]
just merging it cause why not

**Pull request merged**